tests/test_browser.py

```diff
@@ -3299,6 +3300,62 @@ def test_dynamic_link(self):
     # same wasm side module works
     self.btest(self.in_dir('main.cpp'), '2', args=['-s', 'MAIN_MODULE=1', '-O2', '--pre-js', 'pre.js', '-s', 'WASM=1', '-s', 'EXPORT_ALL=1'])
 
+  # verify that dynamic linking works in all kinds of in-browser environments.
+  # don't mix different kinds in a single test.
+  def test_dylink_dso_needed_wasm(self):
+      self._test_dylink_dso_needed(1, 0)
+
+  def test_dylink_dso_needed_wasm_inworker(self):
+      self._test_dylink_dso_needed(1, 1)
+
+  def test_dylink_dso_needed_asmjs(self):
+      self._test_dylink_dso_needed(0, 0)
+
+  def test_dylink_dso_needed_asmjs_inworker(self):
+      self._test_dylink_dso_needed(0, 1)
+
+  @requires_sync_compilation
+  def _test_dylink_dso_needed(self, wasm, inworker):
+    # here we reuse test_core.test_dylink_dso_needed, but the code is run via browser.
+    print('\n# wasm=%d inworker=%d' % (wasm, inworker))
+    self.set_setting('WASM', wasm)
+    self.emcc_args += ['-O2']
+    t = test_core.binaryen2 if wasm else test_core.asm2
```

Comment:
I tried also_asmjs & other options in btest. It doesn't work here (sorry, I'm now too sleepy to remember exactly how it did not work, but for sure I tried hard first with just using btest infrastructure from test_browser).

This part does not really depend on test_core functionality and `t = test_core.binaryen2 if wasm else test_core.asm2` could be dropped - as no settings etc are used from there (setUp and tearDown are not run becuase they cannot without proelyinheriting). I would get t as test_core.TestCoreBase but that symbol is deleted.

What matters here is that all the logic and code from test_dylink_dso_needed is not duplicated, so maybe we should just move it into some shared place, like e.g. test_dylink.py.

Reply:
ok, I will move to shared place and remove dependency on test_core.

Reply:
I general we worry less about duplication in test code than in production code. 

Do we really need this test in test_browser?    I wonder if we have faced this similar problem before where certain tests make sense both in core and in browser?    If you want to create new shared helper code i I think the best place would be `tests/runner.py`.    Files that start with "test_" tend are normally expected to contain test themselves.

Reply:
> I general we worry less about duplication in test code than in production code.

Though I understand the would-be reason here (to concentrate more on the
production part), let me please explain why this harms projects in the long
run. Not caring about tests hygiene makes them messy over time. It starts to be
difficult to add another tests properly and it also starts difficult to make
sure that all tricky details are covered in all duplicate-edited replicas.
In turn this makes effort to do tests properly increasing, and in turn the
consequence is that on average new tests becomes of lesser and lesser quality.
Which in the end hurts the production part because it becomes tested less well.

In this particular case, if I did not push myself hard to make sure
`test_dylink_dso_needed` from core also passes on the browser, and just answered to
request to not change `browser.test_dynamic_link` and simply splitted my original
simpler changes to that test to e.g. `browser.test_dynamic_link_needed`, the
concurrency bug in `loadWebAssemblyModule` would remain undiscovered. And if I just
copy-pasted `core.test_dylink_dso_needed` to test_browser and then adapted it
there to run under `BrowserCore`, over time core and browser versions would
tend to diverge, which would harm making sure that dynamic-linker functionality
is properly tested under both node and on the web.

That's why I tried to avoid duplication.

> Do we really need this test in test_browser?

Yes, synchronous loading does not work in the browser and this way testing ldso
functionality there adds coverage for loadAsync mode. Also the browser is
number-1 target for many web-projects, including my case, so it is better to
have proper coverage for code-paths that are activated only on that
environment.

> I wonder if we have faced this similar problem before where certain tests
> make sense both in core and in browser?

Yes, kind of. I'm not sure about the rest, but e.g. testing how dynamic linking
works in browser, e.g.

2de1bc554e1b00069f92da096fcc74132b2bd278
a5866a5070686295b82a28710e58d8ca3cf637aa
...

could be made to be generic tests and tested under both node and web environments.

In general I would suggest to consider reducing divergence between "regular"
and "in-browser" tests where possible, and to be able to run all, or almost
all, core tests inside browser too.

> If you want to create new shared helper code i I think the best place would
> be tests/runner.py. Files that start with "test_" tend are normally expected to
> contain test themselves.

Ok, I proceed with `tests/runner.py`.

Reply:
> > I general we worry less about duplication in test code than in production code.
> 
> Though I understand the would-be reason here (to concentrate more on the
> production part), let me please explain why this harms projects in the long
> run. Not caring about tests hygiene makes them messy over time. It starts to be
> difficult to add another tests properly and it also starts difficult to make
> sure that all tricky details are covered in all duplicate-edited replicas.
> In turn this makes effort to do tests properly increasing, and in turn the
> consequence is that on average new tests becomes of lesser and lesser quality.
> Which in the end hurts the production part because it becomes tested less well.
> 

Of source, I understand why test hygiene is important.   If you look at the the history over the last year you'll see that I've been doing more than anyone recently to cleanup the existing tests here in emscripten.

My point wasn't that test hygiene isn't important.   Only that the principles we focus on in test code can sometimes be different.  For example, IMHO, DRY can sometimes be less of priority when writing tests.   I might, at times, put readability above DRY in test code.  I this case of course there *is* a lot of case and we wouldn't want to duplicate.  

My objection was mostly to that rather strange inter-class mixing.   

Thanks for refactoring.

Reply:
@sbc100, thanks for feedback. Of course nothing should be taken as absolute. Sometimes it is better to duplicate - e.g. when avoiding the duplication would bring more complexity thus making the whole thing worse overall (but then it applies to production part as well). Just, like you say, in this particular case avoiding the duplication is much better overall.

I agree with the objection and that's why I reworked the patch (hopefully for the better - thanks for review feedback). It was just that I'm completely new to emscripten codebase and was trying to get the test working late at night.

Good to hear that tests are being gardened - thanks for your work on this.